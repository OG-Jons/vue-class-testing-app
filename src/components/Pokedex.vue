<template>
  <h1>Pokedex thing</h1>
  <form @submit.prevent="search">
    <!--    Input field and submit button -->
    <input v-model="searchTerm" type="text" placeholder="Search for a Pokemon" />
    <button type="submit">Search</button>
  </form>
  <p v-if="searching">searching...</p>
  <div v-if="searchResult.pokemon && !searching">
    <p>{{ capitalized }}</p>
    <img :src="searchResult.pokemon.sprites.front_default" alt="Pokemon" />
  </div>
</template>

<script setup lang="ts">
import { reactive, ref, computed } from "vue";
import { AxiosError, AxiosResponse } from "axios";
import axios from "axios";

const searchTerm = ref("");
const searching = ref(false);

const searchResult = reactive({
  pokemon: undefined as { name: string; sprites: { front_default: string } } | undefined,
});

const search = async () => {
  searching.value = true;
  await axios
    ?.get(`https://pokeapi.co/api/v2/pokemon/${searchTerm.value.toLowerCase()}`)
    .then((response: AxiosResponse) => {
      searchResult.pokemon = response.data;
      searching.value = false;
    })
    .catch((e: AxiosError) => {
      searching.value = false;
      if (e.response?.status === 404) {
        searchResult.pokemon = {
          name: "Pokemon not found",
          sprites: {
            front_default: "https://via.placeholder.com/150",
          },
        };
      }
    });
};

const capitalized = computed(
  () => `${searchResult.pokemon?.name.charAt(0).toUpperCase()}${searchResult.pokemon?.name.slice(1)}`,
);
</script>
