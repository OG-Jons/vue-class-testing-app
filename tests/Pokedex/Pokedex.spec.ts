import { mount, VueWrapper } from "@vue/test-utils";
import Pokedex from "../../src/components/Pokedex.vue";
import { expect, vi } from "vitest";
import { testPokemon } from "../setup.vitest";
import axios from "axios";
import { nextTick } from "vue";

describe("Pokedex", () => {
  expect(Pokedex).toBeTruthy();
  const spy = vi.spyOn(axios, "get");
  const wrapper: VueWrapper = mount(Pokedex);

  it("should mount", async () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should search a valid pokemon", async () => {
    spy.mockResolvedValueOnce({
      data: testPokemon,
    });

    const input = wrapper.get("input");
    const button = wrapper.get("button");

    await input.setValue("bulbasaur");
    await button.trigger("submit");

    expect(axios.get).toHaveBeenCalledTimes(1);
    expect(axios.get).toHaveBeenCalledWith("https://pokeapi.co/api/v2/pokemon/bulbasaur");

    await nextTick();

    expect(wrapper.text()).toContain("Bulbasaur");
    expect(wrapper.html()).toContain(testPokemon.sprites.front_default);
  });

  it("should show error on invalid pokemon", async () => {
    spy.mockRejectedValueOnce({
      response: {
        status: 404,
      },
    });

    const input = wrapper.get("input");
    const button = wrapper.get("button");

    await input.setValue("invalid");
    await button.trigger("submit");

    expect(axios.get).toHaveBeenCalledTimes(2);
    expect(axios.get).toHaveBeenCalledWith("https://pokeapi.co/api/v2/pokemon/invalid");

    await nextTick();

    expect(wrapper.text()).toContain("searching...");

    await nextTick();

    expect(wrapper.text()).toContain("Pokemon not found");
    expect(wrapper.html()).toContain("https://via.placeholder.com/150");

    spy.mockClear();
  });
});
