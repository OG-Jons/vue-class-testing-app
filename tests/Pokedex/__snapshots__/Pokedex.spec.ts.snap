// Vitest Snapshot v1

exports[`Pokedex > should mount 1`] = `
"<h1>Pokedex thing</h1>
<form>
  <!--    Input field and submit button --><input type=\\"text\\" placeholder=\\"Search for a Pokemon\\" /><button type=\\"submit\\">Search</button>
</form>
<!--v-if-->
<!--v-if-->"
`;
